### DISCUSSION TITLE
The energy equation add a source term

### DISCUSSION THREAD

**Discussion by czc-zju** [AMReX-Combustion/PeleLMeX#246] [Q&A]
I want to add a source term to the energy equation to achieve ignition, can this be achieved? 
This source term equation:
q=E/(r*t) ( t<tig)
q=0 ( t>tig)

**Comment by drummerdoc** [comment]
Personally, I'd make it a bit smoother than that, using a tanh function in t or something to keep things differentiable and well-behaved.... :)

**Comment by czc-zju** [comment]
Thank you！
In which file should I add this source item?

**Comment by baperry2** [comment]
The `m_extSource` member of the PeleLM class is a vector of (pointers to) multifabs that contains external source terms for all state variables across each level of refinement. The code contains the machinery to handle these terms but right now they are set to 0 everywhere. To add nonzero source terms, you'll have to write your own function that populates `m_extSource` with your desired source terms and use it instead of just setting the value to zero on initialization, when regridding, and when recomputing the external source at each timestep. I believe the following locations in the code should be sufficient:

```
$ grep -rn "m_extSource" | grep setVal
./PeleLMRegrid.cpp:358:   m_extSource[lev]->setVal(0.);
./PeleLMRegrid.cpp:460:   m_extSource[lev]->setVal(0.);
./PeleLMAdvance.cpp:72:     m_extSource[lev]->setVal(0.);
./PeleLMInit.cpp:106:   m_extSource[lev]->setVal(0.);
```

**Reply by baperry2** [reply]
Another quick note on this: in the low-Mach formulation, we transport enthalpy rather than internal energy. If I understand your intent properly, you will want to add a source term for the `RHOH` element of `m_extSource` and have all other elements be set to zero.

**Reply by czc-zju** [reply]
Thank you very much for your response！I am currently trying it out. 
I want added a source term **qig** to the equation of **_h_** (enthalpy), which varies with time and is also dependent on spatial position. The purpose is as follows:
![图片3](https://github.com/AMReX-Combustion/PeleLMeX/assets/81547457/f0d6160c-6298-4a0b-8e76-108fd6757682)
